src/Model/Query.ts

```diff
@@ -111,457 +131,485 @@ Model.prototype._initQueries = function(items) {
+  context: Context;
+  created: boolean;
+  expression: any;
+  extraSegments: string[];
+  fetchCount: number;
+  hash: string;
+  idMap: Record<string, any>;
+  idsSegments: string[];
+  model: Model<ModelData>;
+  options: any;
+  segments: Segments;
+  shareQuery: any | null;
+  subscribeCount: number;
+
+  _pendingSubscribeCallbacks: any[];
+
+  constructor(model: ChildModel<unknown>, collectionName: string, expression: any, options?: any) {
+    // Note that a new childModel based on the root scope is created. Only the
+    // context from the passed in model has an effect
+    this.model = model.root.pass({ $query: this });
+    this.context = model._context;
+    this.collectionName = collectionName;
+    this.expression = util.deepCopy(expression);
+    this.options = options;
+    this.hash = queryHash(this.context.id, collectionName, expression, options);
+    this.segments = ['$queries', this.hash];
+    this.idsSegments = ['$queries', this.hash, 'ids'];
+    this.extraSegments = ['$queries', this.hash, 'extra'];
+
+    this._pendingSubscribeCallbacks = [];
+
+    // These are used to help cleanup appropriately when calling unsubscribe and
+    // unfetch. A query won't be fully cleaned up until unfetch and unsubscribe
+    // are called the same number of times that fetch and subscribe were called.
+    this.subscribeCount = 0;
+    this.fetchCount = 0;
+
+    this.created = false;
     this.shareQuery = null;
-  }
-  this.model.root._queries.remove(this);
-  this.idMap = {};
-  this.model._del(this.segments);
-  this._maybeUnloadDocs(ids);
-};
 
-Query.prototype.fetch = function(cb) {
-  cb = this.model.wrapCallback(cb);
-  this.context.fetchQuery(this);
+    // idMap is checked in maybeUnload to see if the query is currently holding
+    // a reference to an id in its results set. This map is duplicative of the
+    // actual results id list stored in the model, but we are maintaining it,
+    // because otherwise maybeUnload would be looping through the entire results
+    // set of each query on the same collection for every doc checked
+    //
+    // Map of id -> count of ids
+    this.idMap = {};
+  }
 
-  this.fetchCount++;
+  create() {
+    this.created = true;
+    this.model.root._queries.add(this);
+  };
 
-  if (!this.created) this.create();
+  destroy() {
+    var ids = this.getIds();
+    this.created = false;
+    if (this.shareQuery) {
+      this.shareQuery.destroy();
+      this.shareQuery = null;
+    }
+    this.model.root._queries.remove(this);
+    this.idMap = {};
+    this.model._del(this.segments);
+    this._maybeUnloadDocs(ids);
+  };
 
-  var query = this;
-  function fetchCb(err, results, extra) {
-    if (err) return cb(err);
-    query._setExtra(extra);
-    query._setResults(results);
-    cb();
-  }
-  this.model.root.connection.createFetchQuery(
-    this.collectionName,
-    this.expression,
-    this.options,
-    fetchCb
-  );
-  return this;
-};
+  fetch(cb: ErrorCallback) {
```

Comment:
`cb` is technically optional, though in practice the only point in calling w/o callback is if you're doing a throwaway fetch and letting Derby update the UI automatically.

It's optional in the other Query methods, so let's make it optional here too